Array Interface and Categorical internals Refactor

TomAugspurger opened the pull request:
(edit post categorical-move)

Rebased on master. Summary of the changes from master:

    Added the ExtensionArray class
    Categorical subclasses ExtensionArray

    Implements the new methods for the interface (all private. No public API
    changes)

    Adapted the ExtensionDtype class to be the public ABC
    a. Subclass that with PandasExtensionClass that does non-interface things
    like reprs, caching, etc.
    b. All our custom dtypes inherit from PandasExtensionClass, so they implement
    the interface.
    Internals Changes:
    a. Added an ExtensionBlock. This will be a parent for our current custom
    blocks, and the block type for all 3rd-party extension arrays.
    Added a new is_extension_array_dtype method. I think this is nescessary
    for now, until we've handled DatetimeTZ.

This isn't really a test of whether extension arrays work yet, since we're
still using Categorical for everything. I have a followup PR that implements
an IntervalArray that requires additional changes to, e.g., the constructors
so that things work. But all the changes from core/internals.py required to
make that work are present here.

---


1. New class hierarchy in internals

Old:

```python
class CategoricalBlock(NonConsolidatableMixin, ObjectBlock):
    pass
```

new:

```python
class ExtensionBlock(NonConsolidatableMixin, Block):
   pass

class CategoricalBlock(ExtensionBlock):
    pass
```

Figuring out which methods of `ObjectBlock` were required on `CategoricalBlock`
wasn't trivial for me. I probably messed some up.

I think that eventually we can remove `NonConsolidatableMixin`, with the idea
that all non-consolidatable blocks are blocks for extension dtypes? That's true
today anyway.

Followup PRs:

1. Making `core/arrays/period.py` and refactoring `PeriodIndex`
2. Making `core/arrays/interval.py` and refactoring `IntervalIndex`
3. Adding docs and generic tests like https://github.com/pandas-dev/pandas/pull/19174/files#diff-e448fe09dbe8aed468d89a4c90e65cff for our interface (once it's stabilized a bit).

TomAugspurger committed:
REF: Define extension base classes

pep8speaks commented:
Hello @TomAugspurger! Thanks for submitting the PR.







 - In the file [`pandas/core/arrays/categorical.py`](https://github.com/pandas-dev/pandas/blob/6e084ad9447518f418c28448fdadd65e52f6a85c/pandas/core/arrays/categorical.py), following are the PEP8 issues :

> [Line 560:9](https://github.com/pandas-dev/pandas/blob/6e084ad9447518f418c28448fdadd65e52f6a85c/pandas/core/arrays/categorical.py#L560): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'

















 - In the file [`pandas/core/internals.py`](https://github.com/pandas-dev/pandas/blob/6e084ad9447518f418c28448fdadd65e52f6a85c/pandas/core/internals.py), following are the PEP8 issues :

> [Line 119:80](https://github.com/pandas-dev/pandas/blob/6e084ad9447518f418c28448fdadd65e52f6a85c/pandas/core/internals.py#L119): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 1711:80](https://github.com/pandas-dev/pandas/blob/6e084ad9447518f418c28448fdadd65e52f6a85c/pandas/core/internals.py#L1711): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,187 @@
+            Should return an ExtensionArray, even if ``self[slicer]``
+            would return a scalar.
+        """
+        # XXX: We could get rid of this *if* we require that
```
We can *almost* get rid of `_slice` with a default implementation noted in the comments, but I think dimensionality reduction from array -> scalar could break things.

jorisvandenbossche replied:
Assuming `slicer` is a slice, in what cases can it give a scalar?

TomAugspurger replied:
`slice(0)` hits this.

Ahh, but I didn't realize that NumPy always returned arrays from slices:

```python
In [5]: np.array([0])[slice(0, 1)]
Out[5]: array([0])
```

So if we assume that as well, then we could get rid of `_slice`.

jorisvandenbossche replied:
Yes, I think we can assume that if `__getitem__` gets a `slice` object, it should return an instance of itself.

TomAugspurger commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,89 @@
+    @property
+    def type(self):
+        """Typically a metaclass inheriting from 'type' with no methods."""
+        return type(self.name, (), {})
```
I'm not really sure what `array.dtype.type` is used for. This passes the test suite, but may break things like

```
array1.dtype.type is array2.dtype.type
```

since the object IDs will be different (I think).

shoyer replied:
In NumPy, `dtype.type` is the corresponding scalar type, e.g.,
```
>>> np.dtype(np.float64).type
numpy.float64
```

I don't know where "Typically a metaclass inheriting from 'type' with no methods." comes from.

TomAugspurger replied:
Thanks, that makes sense. Would you say that `object` is a good default here? I'll work through a test array that uses something meaningful like `numbers.Real` or `int`. We do use `values.dtype.type` in a few places, like figuring out which block type to use.

shoyer replied:
I don't think there's a good default value here. Generally the right choice is to return the corresponding scalar type, e.g., `Interval` for `IntervalDtype`.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -108,14 +109,15 @@ class Block(PandasObject):
         if ndim is None:
             ndim = values.ndim
-        elif values.ndim != ndim:
+        elif self._validate_ndim and values.ndim != ndim:
```
@jreback would be curious to hear your thoughts here. Needed this so that `ExtensionBlock` could inherit from `Block` and call `super().__init__`. We *could* avoid this by setting `.values`, `.mgr_locs`, `.ndim` directly in `ExtensionBlock.__init__`, but I think it's best practice to always call your parent's init.

jreback replied:
yes, see my above comment

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -1697,24 +1698,20 @@ class NonConsolidatableMixIn(object):
+        # Placement must be converted to BlockPlacement so that we can check
+        # its length
+        if not isinstance(placement, BlockPlacement):
+            placement = BlockPlacement(placement)
```
I found it much clearer to just copy these two lines rather than calling `mgr_locs.setter`, but can revert to that if desired.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -2360,23 +2443,13 @@ def is_view(self):
     def to_dense(self):
         return self.values.to_dense().view()
 
-    def convert(self, copy=True, **kwargs):
```
In the ExtensionBlock now.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -2360,23 +2443,13 @@ def is_view(self):
         """
         return np.object_
 
-    def _slice(self, slicer):
```
In ExtensionBlock now.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -2468,7 +2541,8 @@ class DatetimeBlock(DatetimeLikeBlockMixin, Block):
     def __init__(self, values, placement, fastpath=False, **kwargs):
-        if values.dtype != _NS_DTYPE:
+        if values.dtype != _NS_DTYPE and values.dtype.base != _NS_DTYPE:
+            # not datetime64 or datetime64tz
```
This was so `NonConsolidatableMixin` could call `__init__`. I *think* it's harmless.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -1823,6 +1820,91 @@ def _unstack(self, unstacker_func, new_columns):
+
+    # Methods we can (probably) ignore and just use Block's:
+
+    # * replace / replace_single
```
Any things on this @jreback? CategoricalBlock.replace used to be `ObjectBlock.replace`, but now it's just `Block.replace`, which is much simpler.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -1823,6 +1820,91 @@ def _unstack(self, unstacker_func, new_columns):
+            values = values.reshape((1,) + values.shape)
+        return values
+
+    def _can_hold_element(self, element):
```
And @jreback I'll defer to you on this one, since you know more about when this is called. I don't think we can just do `return isinstance(element, self._holder)`, since `element` may need to be coerced. For an IP address, we want

```python
ser = pd.Series(ip.IPAddress(['192.168.1.1']))
ser[0] = '192.168.1.0'
```
to work, and (I think) `element` would just be the str there.

jorisvandenbossche commented:
@TomAugspurger for reviewing, did you need to change a lot to `Categorical` implementation itself? (because you moved it hard to see ..)

jreback commented:
i would much prefer that moves occur separately from changes

TomAugspurger commented:
I did keep the move in https://github.com/pandas-dev/pandas/commit/4b06ae4338adfa6efb47d02a95362a82e329e55e

All the non-move changes to categorical are in https://github.com/pandas-dev/pandas/pull/19268/commits/a9e097228fcd48d7cbb0a1ef6894bd805f5f84c2#diff-f3b2ea15ba728b55cab4a1acd97d996d

TomAugspurger commented:
But I can split the move into its own PR as well.

One thing: do we have code for pickle compat on module changes? The old pickle files expect a `pandas.core..categorical.Categorical`. I threw a shim in there, but would be nice to remove it.

TomAugspurger commented:
Splitting these PRs now (I'll make a new PR for the move and then rebase this one)

TomAugspurger commented:
https://github.com/pandas-dev/pandas/pull/19269 for the move.

TomAugspurger commented:
Rebased on master. Summary of the changes from master:

1. Added the ExtensionArray class
2. `Categorical` subclasses `ExtensionArray`
  - Implements the new methods for the interface (all private. No public API
    changes)
3. Adapted the `ExtensionDtype` class to be the public ABC
  a. Subclass that with `PandasExtensionClass` that does non-interface things
     like reprs, caching, etc.
  b. All our custom dtypes inherit from `PandasExtensionClass`, so they implement
     the interface.
4. Internals Changes:
   a. Added an `ExtensionBlock`. This will be a parent for our current custom
      blocks, and the block type for all 3rd-party extension arrays.
5. Added a new `is_extension_array_dtype` method. I think this is nescessary
   for now, until we've handled DatetimeTZ.

This isn't *really* a test of whether extension arrays work yet, since we're
still using `Categorical` for everything. I have a followup PR that implements
an `IntervalArray` that requires additional changes to, e.g., the constructors
so that things work. But all the changes from `core/internals.py` required to
make that work are present here.

TomAugspurger commented:
@shoyer, @chris-b1, @wesm if you're interested in the interface, this is the PR it's being defined in, though we may have to make refinements later.

I have followup PRs ready for `IntervalArray` and `PeriodArray` that build on this. Docs and such will come in the `IntervalArray` PR, once we have publicly visible changes.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+    # ------------------------------------------------------------------------
+    @property
+    def base(self):
+        """The base array I am a view of. None by default."""
```
Can you give an example here?

Perhaps it would also help to explain how is this used by pandas?

TomAugspurger replied:
It's used in `Block.is_view`, which AFAICT is only used for chained assignment?

If that's correct, then I think we're OK with saying this purely for compatibility with NumPy arrays, and has no effect. I've currently defined `ExtensionArray.is_view` to always be `False`, so I don't even make use of it in the changes so far.

shoyer replied:
I would also remove this. NumPy doesn't always maintain this properly, so it can't actually be essential.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+    @abc.abstractmethod
+    def take(self, indexer, allow_fill=True, fill_value=None):
+        # type: (Sequence, bool, Optional[Any]) -> ExtensionArray
+        """For slicing"""
```
This should clarify what valid values of `indexer` are. Does -1 indicate a fill value?

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+
+    def take_nd(self, indexer, allow_fill=True, fill_value=None):
+        """For slicing"""
+        # TODO: this isn't really nescessary for 1-D
```
I would remove this

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+
+    @property
+    def is_sparse(self):
+        """Whether your array is sparse. True by default."""
```
Correction: False by default :)

This should clarify what it means to be a sparse array. How does pandas treat sparse arrays differently?

I would consider dropping this if it isn't strictly necessary.

TomAugspurger replied:
I think it's unnecessary.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+            Should return an ExtensionArray, even if ``self[slicer]``
+            would return a scalar.
+        """
+        return type(self)(self[slicer])
```
This default implementation is likely to fail for some obvious implementations. Perhaps we can have a constructor method `_from_scalar()` instead that converts a scalar into a length 1 array?

TomAugspurger replied:
Let me see if I can verify that this is always called with a `slice` object. In that case, `__getitem__` will return an `ExtensionArray`, and we don't have to worry about the scalar case. Unless I'm missing something.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        from pandas.core.algorithms import value_counts
+        mask = ~np.asarray(self.isna())
+        values = self[mask]  # XXX: this imposes boolean indexing
```
We should have a dedicated method for boolean indexing or document this as part of the expected interface for `__getitem__`.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+    @abc.abstractmethod
+    def get_values(self):
+        # type: () -> np.ndarray
+        """Get the underlying values backing your data
```
This isn't very clear. How does this differ from `.base`?

TomAugspurger replied:
IIUC, `.base` is used to get the base array `self` is a view of (which will likely be `None` for our purposes?).

`.get_values()` is to convert `self` to an `ndarray`. This may be useful if `np.asarray(self)`, which would fall back to `self.__iter__`, doesn't do the right thing.

I'll try to give some examples of where this is used.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+
+
+@add_metaclass(abc.ABCMeta)
+class ExtensionArray(object):
```
Are there any expected requirements for the constructor `__init__`?

TomAugspurger replied:
We also expect that `ExtensionArray()`, with no arguments, works so that subclasses don't have to implement `construct_from_string`.

Rather than imposing that on subclasses, we could require some kind of `.empty` alternative constructor.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+    def dtype(self):
+        """An instance of 'ExtensionDtype'."""
+        # type: () -> ExtensionDtype
+        pass
```
Please drop `pass` from all these methods. It's not needed (docstrings alone suffice).

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,201 @@
+    @abc.abstractmethod
+    def nbytes(self):
+        """The number of bytes needed to store this object in memory."""
+        # type: () -> int
```
Type comments come before the docstring: http://mypy.readthedocs.io/en/latest/python2.html

shoyer commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,92 @@
+
+    @property
+    def kind(self):
+        """A character code (one of 'biufcmMOSUV'), default 'O'
```
This should clarify how it's used. How is this useful?

Perhaps "This should match `dtype.kind` when arrays with this dtype are cast to numpy arrays"?

TomAugspurger committed:
Updated for comments

* removed take_nd
* Changed to_dense to return get_values
* Fixed docstrings, types
* Removed is_sparse

TomAugspurger committed:
Remove metaclasses from PeriodDtype and IntervalDtype

TomAugspurger committed:
Fixup form_blocks rebase

TomAugspurger committed:
Restore concat casting cat -> object

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -5,15 +5,15 @@
 
-class ExtensionDtype(object):
+
+class PandasExtensionDtype(ExtensionDtype):
```
why would this not be just ExtensionDtype

TomAugspurger replied:
We define methods like `__repr__`, caching, etc. that are not part of the interface.

jreback replied:
__repr__ is reasonable to be part of the interface, caching I suppose is ok here

jreback commented on pandas/core/internals.py:
```diff
@@ -95,6 +96,7 @@ class Block(PandasObject):
     is_object = False
     is_categorical = False
     is_sparse = False
+    is_extension = False
```
you wouldn't do it like this. rather you would inherit from an ExtensionBlock

TomAugspurger replied:
I haven't looked at what these `is_` properties are used for, but all the other blocks had them so I added it for consistency.

jreback replied:
let's not add things like this until / unless necessary

jreback commented on pandas/core/internals.py:
```diff
@@ -1821,6 +1824,130 @@ def _unstack(self, unstacker_func, new_columns):
         return blocks, mask
 
 
+class ExtensionBlock(NonConsolidatableMixIn, Block):
```
this should NOT be in this file, make a dedicated namespace. pandas.core.internals.block or something

TomAugspurger replied:
Why would it not be in this file? All the other blocks are. Note that `ExtensionBlock` is not ever going to be public, only `ExtensionArray` and `ExtensionDtype`.

jreback replied:
this is exactly the point, this diff is already way too big. let's do a pre-cursor PR to split Block and manger (IOW internals into 2 files).

jorisvandenbossche replied:
By splitting the file first, the diff here will not be smaller

jorisvandenbossche commented on pandas/core/internals.py:
```diff
@@ -1821,6 +1824,130 @@ def _unstack(self, unstacker_func, new_columns):
+        # ExtensionArrays must be iterable, so this works.
+        values = np.asarray(self.values)
+        if values.ndim == self.ndim - 1:
+            values = values.reshape((1,) + values.shape)
```
Is this needed? I know it currently like that in `NonConsolidatableBlock`, but do we ever expect the result to be a 2D array if this is holded in an DataFrame. 
Eg datetimetz block returns here a DatetimeIndex for both `.values` and `.get_values()`. On the other hand, a categorical block does this reshaping and returns Categorical vs 2d object numpy array.

Further, do we need to do something with `dtype` arg?

jreback replied:
this is for sparse. which is a step-child ATM. has to be dealt with

jorisvandenbossche reviewed (commented):
I will try to test this out with GeoPandas one of the coming days, to give some more feedback

jreback commented:
> I think that eventually we can remove NonConsolidatableMixin, with the idea
that all non-consolidatable blocks are blocks for extension dtypes? That's true
today anyway.

yes you should folk NonConsolidatedBlock into the ExtensionBlock infrastructure. maybe be slightly tricky right now as Sparse has some special casing.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -149,7 +151,7 @@ def _maybe_to_categorical(array):
 
 
-class Categorical(PandasObject):
+class Categorical(ExtensionArray, PandasObject):
```
yeah, the methods in PandasObject needs to be ABC in the ExtensionArray

jorisvandenbossche replied:
On the other hand, I don't think all methods/attributes of PandasObject should be added to the public ExtensionArray (to keep those internal + to not clutter the ExtensionArray API)

TomAugspurger replied:
FYI, I'm consistently testing these changes against

1. An implementation of `IntervalArary`: https://github.com/TomAugspurger/pandas/compare/pandas-array-interface-3...TomAugspurger:pandas-array-upstream+interval?expand=1
2. A branch on pandas-ip: https://github.com/ContinuumIO/pandas-ip/tree/pandas-array-upstream-compat

Neither inherit from `PandasObject` at the moment, so we're OK.

jreback commented on pandas/core/internals.py:
```diff
@@ -2410,29 +2525,6 @@ def shift(self, periods, axis=0, mgr=None):
                                           placement=self.mgr_locs)
 
-    def take_nd(self, indexer, axis=0, new_mgr_locs=None, fill_tuple=None):
-        """
```
so as I said above, this PR is doing way way too much. pls just create the ExtensionBlock and just move things. Then in another PR you can do the changes.

jorisvandenbossche replied:
This method is just moved a bit up in the file (to the parent class) ?

TomAugspurger replied:
Yep.

TomAugspurger replied:
This really is doing the bare minimum changes to `core/internals.py`. This won't even allow, e.g. `pd.Series(MyExtensionArray())` to work, yet, though I have a follow PR ready to go that does that (with `IntervalArray` as a test case).

TomAugspurger committed:
Remove _slice, clarify semantics around __getitem__

jbrockmendel commented:
> this is exactly the point, this diff is already way too big. let's do a pre-cursor PR to split Block and manger (IOW internals into 2 files).

 Some overnight-rebasing notwithstanding, I've got a branch ready that splits `internals` into `internals.blocks`, `internals.managers`, `internals.joins` (and except for import updates is pure cut/paste).  Would it be helpful to push that now?  I had planned to wait because it will make a ton of rebasing necessary in other PRs.

shoyer commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,101 @@
+    @abc.abstractmethod
+    def type(self):
+        # type: () -> type
+        """The scalar type for your array, e.g. ``int`` or ``object``."""
```
Let's not encourage using `object` for user defined types :)

shoyer commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,101 @@
+        """A character code (one of 'biufcmMOSUV'), default 'O'
+
+        This should match the NumPy dtype used when your array is
+        converted to an ndarray, which is probably 'O' for object.
```
Clarify: (if your data type does not corespond to a built-in numpy dtype)

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,194 @@
+
+        * Your class should be able to be constructed with no arguments,
+          i.e. ``ExtensionArray()`` returns an instance.
+          TODO: See comment in ``ExtensionDtype.construct_from_string``
```
What exactly is `ExtensionArray()` supposed to return? A length zero array? This should be clarified.

I don't understand the comment about `construct_from_string`. `construct_from_sting` currently returns an instance of the dtype object, not an array.

Also, in my opinion is somewhat poor API design for an empty constructor to automatically correspond to a certain result. I would rather have an explicit `empty()` class method.

TomAugspurger replied:
Ignore my comments about `ExtensionArray()` . That was apparently vestigial from an earlier version of `construct_from_string`. Agreed that this would have been an awful design decision to force on subclasses :)

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,194 @@
+    @abc.abstractmethod
+    def nbytes(self):
+        # type: () -> int
+        """The number of bytes needed to store this object in memory."""
```
What should a user do if this is expensive or otherwise difficult to calculate properly? For example, if it's a numpy array with `dtype=object`.

We should probably note that it's OK for this to be an approximate answer (a lower bound) on the number of required bytes, and consider adding another method for the benefit of `memory_usage()`.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,194 @@
+    @abc.abstractmethod
+    def get_values(self):
+        # type: () -> np.ndarray
+        """A NumPy array representing your data.
```
Let's clarify:
- This is used for `Series.values`, but otherwise is not used internally by pandas (This is true, right?).
- If I've implemented a custom scalar type, should this be an object array of containing all elements of that type? Or should I prefer something possibly lossy but faster? e.g., for IntervalArray, should I return an object array of Interval objects, or a structured array with two elements? It's OK if both options are fine depending on the circumstances, let's just give a little guidance.

TomAugspurger replied:
This is *not* used by `Series.values`, that currently returns the array, like `pd.Series(categorical)` is a `Categorical`. It seems to be used in

1. indexing, via `_values_from_object`, which seems undesirable
2. nanops, via `_values_from_object`, again undesirable

So let's hold off on discussing this until I can understand it better. We could potentially remove it. And if extension arrays are iterable, then we could just as well use `np.array(extension_array, dtype='object')` whenever we absolutely need an ndarray of objects.

jorisvandenbossche replied:
It is also used for eg concatting that ends up in object dtype, eg in case of concatting different types (a rather corner case). 
And it is also similar to `Series.get_values`.

I *think* in general this is the "fallback numpy array" in case we need one and we need to be able to infer what to do with it (so typically the object array of custom scalars).

It could be an option to use this to enable (or enable good error message) operations. For example that those values are used when the user calls for the `sum`. And if you then have an object array of custom scalars, this will be handled gracefully. 
Although (thinking while I am writing), this is in general also not ideal as we would want: a) the ability to let pandas operations dispatch to efficient implementations of the array itself (eg in hypothetical case you have a `sum` function on your custom array that does not need to construct object array of scalars) and b) the ability to disable numerical operations without the need to crystallize a full object array (as eg in geopandas this is very costly)

TomAugspurger replied:
> I think in general this is the "fallback numpy array" in case we need one and we need to be able to infer what to do with it (so typically the object array of custom scalars).

Yep. That sounds right.

> let pandas operations dispatch to efficient implementations of the array itself [...] and b) the ability to disable numerical operations without the need to crystallize a full object array

Agreed on both accounts, though I think getting reductions, cumulative ops, and numeric ops to all work properly on extension arrays will require some careful thought an quite a bit of work. At the moment, I'm hoping to just avoid *breaking* things like `df.mean()` when you have a mix of numeric and extension columns.

TomAugspurger replied:
See https://github.com/pandas-dev/pandas/issues/19318 for notes about `get_values` being called during indexing, specifically `Series.__getitem__`. I think it can be avoided.

shoyer commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,101 @@
+        if string == cls.name:
+            # XXX: Better to mandate a ``.from_empty`` classmethod
+            # rather than imposing this on the constructor?
+            return cls()
```
At the very least, this requirement for the constructor should be documented.

shoyer replied:
We still need to document this.

TomAugspurger replied:
Perhaps it's best to just remove the default implementation and make it abstract? I'll document current implementation as a possible default.

TomAugspurger committed:
Document and use take.

TomAugspurger committed:
Clarify type, kind, init

TomAugspurger committed:
Remove base

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,194 @@
+        if the slice is length 0 or 1.
+
+        For scalar ``key``, you may return a scalar suitable for your type.
+        The scalar need not be an instance or subclass of your array type.
```
I'll clarify this to say

        For scalar ``item``, you should return a scalar value suitable
        for your type. This should be an instance of ``self.dtype.type``.

My earlier phrasing was to explain that the return value for scalars needn't be the type of item that's actually stored in your array. E.g. for my IPAddress example, the array holds two uint64s, but a scalar slice returns an `ipaddress.IPv4Address` instance.

TomAugspurger commented on pandas/core/dtypes/common.py:
```diff
@@ -1685,6 +1685,38 @@ def is_extension_type(arr):
+
+    # we want to unpack series, anything else?
+    if isinstance(arr_or_dtype, ABCSeries):
+        arr_or_dtype = arr_or_dtype.values
```
This is `Series.values`, what else would it return? An object-typed NumPy array? I think the ship has sailed on `Series.values` always being a NumPy array.

jorisvandenbossche replied:
Let's use `Series._values` for now? That gets the values of the block, and is certainly an ExtensionArray in case the series holds one. 
The we can postpone the decision on what `.values` returns?

TomAugspurger commented:
Thanks for the comments, tried to address most inline, going back through now to see if I missed any.

> Some overnight-rebasing notwithstanding, I've got a branch ready that splits internals into internals.blocks, internals.managers, internals.joins (and except for import updates is pure cut/paste). Would it be helpful to push that now? I had planned to wait because it will make a ton of rebasing necessary in other PRs.

Doesn't matter for me. I'm only touching `Blocks` so the git conflicts shouldn't be terrible. I don't recall what else is outstanding. Either way, the diff here will be identical.

jbrockmendel commented:
> Agreed on both accounts, though I think getting reductions, cumulative ops, and numeric ops to all work properly on extension arrays will require some careful thought an quite a bit of work

This is the bit that most closely matches my comparative advantage.  The upshot is that we can adapt the existing FooIndex implementations in most cases, but need to be careful not to let the set of Array methods transferred from Index snowball.

jorisvandenbossche commented:
I quickly tried this branch with geopandas, and it turns out the `GeometryArray` we have there already almost entirely satisfies the interface (apart from the dtype and some of the block related methods). 
I could also put this in a Series and do some basic operations (get item, equality comparison, repr, concat)

First things I noticed:

- I think `__iter__` is not strictly needed to be included in the interface (if you have len and getitem, it will already be iterable)
- `s.isna()` is not working for me (but not sure if it was already meant to be working)
- `align` results in an ObjectBlock (and also eg `pd.merge`, but that is probably related). `align` is something that is working with my GeometryBlock, so I would need to check which part of it is that is making this work compared to ExtensionBlock.
- the `ExtensionDtype.name` is not used in eg `df.dtypes` (although it is used in the series repr). But given that, I think it would be good to have a default `__repr__` implementation that just returns the `name` ?

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,203 @@
+"""An interface for extending pandas with custom arrays."""
```
I'd advocate leaving base.py open for (near-)future usage as pandas-internal base and putting the "use this if you want to write your own" file in e.g. extension.py

TomAugspurger replied:
I have a slight preference for `base.py` since it's a base class for all extension arrays. I don't think that having `ExtensionArray` in `arrays.base`  precludes having a pandas-internal base there as well.

jorisvandenbossche replied:
It will be publicly exposed through pd.api.extensions anyway I think

TomAugspurger replied:
Adding stuff to the public API is waiting on https://github.com/pandas-dev/pandas/pull/19304

TomAugspurger commented:
Thanks for trying it out @jorisvandenbossche, that's encouraging.

> s.isna() is not working for me (but not sure if it was already meant to be working)

That'll require https://github.com/TomAugspurger/pandas/compare/pandas-array-interface-3...TomAugspurger:pandas-array-upstream+interval#diff-ff8364cee9a3e1ef3a3825cb2cdd26d8 (that's my interval PR). Personally, I think it makes sense to move those changes to this PR, but I'm trying to keep things in small chunks.

TomAugspurger committed:
API: Remove unused __iter__ and get_values

TomAugspurger committed:
API: Implement repr and str

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pandas-array-interface-3

jorisvandenbossche commented:
@jreback what are your thoughts here on the way forward? I would rather try to merge this quickly, with the idea to do follow-up PRs. 
(the problem is that this PR will never be perfect, unless it does at once a full conversion of the other internal types and edits throughout the code-base to use this, but that makes it then a huge diff that is impossible to review)

TomAugspurger committed:
Remove default value_counts for now

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,175 @@
+        ``item`` may be one of
+
+            * A scalar integer position
+            * A slice object
```
Clarify: where start, stop and step are either integers or None

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,175 @@
+
+            * A scalar integer position
+            * A slice object
+            * A boolean mask the same length as 'self'
```
Clarify: where mask is a 1d numpy array?

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,175 @@
+            will be done. This short-circuits computation of a mask. Result is
+            undefined if allow_fill == False and -1 is present in indexer.
+        fill_value : any, default None
+            Fill value to replace -1 values with
```
Add: By default, should use the missing fill value for this type (`_fill_value`).

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,175 @@
+        -----
+        This should follow pandas' semantics where -1 indicates missing values.
+        Positions where indexer is ``-1`` should be filled with the missing
+        value for this type.
```
Let's strongly consider exposing a helper function to make this easier to write, or at least an example of what this would look like (we can save this for later). It's not obvious how to write this with NumPy.

TomAugspurger replied:
I added an example that I hope gets things correct for an extension type backed by a NumPy structured array.

One trouble with this providing a helper function is that we don't know much about how the extension array is actually storing the data. Although, we could rely on the assumption that the underlying storage is convertible to a NumPy array, and proceed from there. Though this would perhaps be sub-optimal for many extension arrays.

jorisvandenbossche replied:
For GeometryArray, we have followed the same idea the write `take` (so it's not necessarily only if you have a *structured* array, just when you have an array backing up your ExtensionArray)

jreback replied:
this is fine as it is a pandas standard.

shoyer reviewed (commented):
I have a few minor comments on the API, but this is looking pretty good now!

TomAugspurger committed:
Fixed merge conflicts

TomAugspurger committed:
Remove implementation of construct_from_string

TomAugspurger committed:
Example implementation of take

TomAugspurger commented:
FYI, I'll be without internet, but working on this, for most of tomorrow.

TomAugspurger committed:
Cleanup ExtensionBlock

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pandas-array-interface-3

TomAugspurger committed:
Pass through ndim

jreback commented:
will have a look later today

jorisvandenbossche commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,122 @@
+    """
+
+    def __repr__(self):
+        return str(self)
```
Something like `return "{}()".format(self.__class__.__name__)` ?

Or we could also remove this. The default will be just a longer version from what I propose.

TomAugspurger replied:
Perhaps remove for now, and do something like scikit-learn does with building the repr from `__init__`'s signature in the future?

TomAugspurger committed:
Use series._values

TomAugspurger committed:
Removed repr, updated take doc

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+        """
+
+    def __setitem__(self, key, value):
+        # type: (Any, Any) -> None
```
setitem *must* be defined (it certainly does not need to actually set inplace), but since we use this is must appear as a mutable object

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+    @property
+    @abc.abstractmethod
+    def dtype(self):
+        # type: () -> ExtensionDtype
```
examples and document

TomAugspurger replied:
What would an example even look like here? Defining an `ExtensionDtype` in the docstring?

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+    def ndim(self):
+        # type: () -> int
+        """Extension Arrays are only allowed to be 1-dimensional."""
+        return 1
```
this should be tested on registration of the sub-type

TomAugspurger replied:
What do you mean "registration"? We could override `ABC.register`, but I don't think there's an (easy) way to validate this if they just subclass `ExtensionArray`.

If people want to mess with this, that's fine, their stuff just won't work with pandas.

jreback replied:
what I mean is that when you register things, we should actually test that the interface is respected. If we had final methods this would not be necessary, but if someone override ndim this is a problem.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+    @abc.abstractmethod
+    def copy(self, deep=False):
+        # type: (bool) -> ExtensionArray
+        """Return a copy of the array."""
```
document.

TomAugspurger committed:
Various cleanups

jreback commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,119 @@
+    def type(self):
+        # type: () -> type
+        """The scalar type for your array, e.g. ``int``
+
```
document more on these

jreback commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,119 @@
+
+    @property
+    def names(self):
+        # type: () -> Optional[List[str]]
```
what is this for?

TomAugspurger replied:
We use it in `DataFrame.__init__`

```python
        elif isinstance(data, (np.ndarray, Series, Index)):
           if data.dtype.names:
```

I could of course modify that, but perhaps as a followup. For now having a non-abstract names property seems OK.

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -1685,6 +1685,38 @@ def is_extension_type(arr):
+    # we want to unpack series, anything else?
+    if isinstance(arr_or_dtype, ABCSeries):
+        arr_or_dtype = arr_or_dtype._values
+    return isinstance(arr_or_dtype, (ExtensionDtype, ExtensionArray))
```
you need to call ``_get_dtype_type`` here, this can *only* have a result of ExtensionDtype and NOT ExtensionArray, which doesn't make any sense.

TomAugspurger replied:
> this can only have a result of ExtensionDtype and NOT ExtensionArray, which doesn't make any sense.

The result is just True or False. The argument can be either an array or dtype.

I'm not sure that `_get_dtype_or_type` does what we want here. That grabs `arr.dtype.type`, which is a scalar like `str` or `Interval` or `ipaddress.IPv4Address`. What would we do with that?

jreback replied:
this is not following what we do in all other cases that's my point. pls use _get_dtype_or_type

TomAugspurger replied:
> this is not following what we do in all other cases that's my point.

That won't work.

```
In [1]: import pandas as pd; import pandas_ip as ip

In [2]: arr = ip.IPAddress([1, 2, 3])

In [3]: pd.core.dtypes.common._get_dtype_type(arr)
Out[3]: pandas_ip.block.IPv4v6Base
```

`IPv4v6Base` isn't an instance of `ExtensionType`. It's the type scalars belong to.

```python
In [4]: isinstance(arr[0], ip.block.IPv4v6Base)
Out[4]: True

In [5]: issubclass(ip.block.IPv4v6Base, pd.core.dtypes.base.ExtensionDtype)
Out[5]: False
```

`_get_dtype_or_type` works for our extension types, since if we get a `CategoricalDtypeType` we can say "this is a categorical". But we can't do that for arbitrary 3rd party types.

jreback replied:
then ``_get_dtype_or_type`` needs adjustment. This is the point of compatibility, there shouldn't be the need to have special cases.

jorisvandenbossche replied:
But that would make `get_dtype_or_type` inconsistent, as it would no longer *always* return a dtype type, but in certain cases a dtype

TomAugspurger replied:
`_get_dtype_type` does exactly what it's supposed to do, `values.dtype.type`. But that's not useful here!

What's the issue with the function as defined? I need a way to tell if an array or dtype is an ExtensionArray or ExtensionDtype. Someday, when `Categorical`, `SparseArray`, `IntervalArray`, `PeriodArray`, `datetimetz`, etc are extension arrays then all the *special cases* currently in `_get_dtype_type` and friends can be removed, but we aren't there yet. We're doing things in small steps.

jreback commented on pandas/core/internals.py:
```diff
@@ -109,14 +111,15 @@ class Block(PandasObject):
         if ndim is None:
             ndim = values.ndim
-        elif values.ndim != ndim:
+        elif self._validate_ndim and values.ndim != ndim:
```
make a method ``_validate_ndim(ndim)`` to do this

TomAugspurger replied:
Hmm, I'd find it surprising to call a method like `self._validate_ndim` and have it not actually do validation. We could call it `_maybe_validate_ndim`, but then you have to know why it's 'maybe'...

jreback replied:
my point is its much better to have these as methods and not properties

TomAugspurger replied:
I'm probably misunderstanding. https://github.com/pandas-dev/pandas/pull/19268/commits/9c06b13d2c8ecf53d678a0ec613acea927bb3955 moved that check to a method, but I don't think it's clearer.

Note that `NonConsolidatableMixin` already had the `_validate_ndim` property. The only reason we need to add the `if self._validate_ndim` is since all child blocks now eventually call `Block.__init__`. The whole `_validate_ndim` stuff seems like a code smell to be anyway, but that cleanup needn't be done in this PR I think.

jreback commented on pandas/core/internals.py:
```diff
@@ -1795,6 +1797,99 @@ def _unstack(self, unstacker_func, new_columns):
+    ExtensionArrays are limited to 1-D.
+    """
+    is_extension = True
+    is_bool = False
```
why are you adding is_bool?

TomAugspurger replied:
Thanks. Removed.

jreback commented on pandas/core/internals.py:
```diff
@@ -1795,6 +1797,99 @@ def _unstack(self, unstacker_func, new_columns):
+    is_bool = False
+
+    def __init__(self, values, placement, ndim=None):
+        self._holder = type(values)
```
might need some assertions here about values, IOW it must be a sub-class of ExtensionArray?

jreback commented on pandas/core/internals.py:
```diff
@@ -1795,6 +1797,99 @@ def _unstack(self, unstacker_func, new_columns):
+        return self.values.get_values()
+
+    def take_nd(self, indexer, axis=0, new_mgr_locs=None, fill_tuple=None):
+        """
```
add a doc-string

jreback commented on pandas/core/internals.py:
```diff
@@ -1795,6 +1797,99 @@ def _unstack(self, unstacker_func, new_columns):
+        else:
+            fill_value = fill_tuple[0]
+
+        # axis doesn't matter; we are really a single-dim object
```
huh?

jorisvandenbossche replied:
this is copy pasted from existing code, and is explaining why the `axis` argument is ignored

jreback replied:
a doc-string would help here

jreback commented on pandas/core/internals.py:
```diff
@@ -1795,6 +1797,99 @@ def _unstack(self, unstacker_func, new_columns):
+        # this onto the array.
+        return True
+
+    def convert(self, copy=True, **kwargs):
```
let' start much more minimally and add methods individually, rather than the reverse (of adding everything).

jorisvandenbossche replied:
This is copy pasted from CategoricalBlock, which now inherits from this one. So we need to keep it in either places to not change anything. You mean keep it in CategoricalBlock?

But, I think this one is actually the same as base Block, and if we leave it out here is just inherited from there. So think we can remove it.

TomAugspurger replied:
Removed.

jreback commented on pandas/core/internals.py:
```diff
@@ -2437,7 +2507,8 @@ class DatetimeBlock(DatetimeLikeBlockMixin, Block):
 
     def __init__(self, values, placement, ndim=None):
-        if values.dtype != _NS_DTYPE:
+        if values.dtype != _NS_DTYPE and values.dtype.base != _NS_DTYPE:
```
huh?

TomAugspurger replied:
Does the comment immediately below explain it?

TomAugspurger replied:
If your question was about why the change was needed, that's because `NonConsolidatableMixin` now calls `super().__init__`, so `DatetimeTZBlock.__init__` ends up here.

jreback replied:
my qestion is why are you accesing base, it should not be needed (so why the change)

TomAugspurger replied:
(without the change)

```python
In [2]: pd.Series(pd.date_range("2017", periods=4, tz="US/Central"))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-bafaac91d0ed> in <module>()
----> 1 pd.Series(pd.date_range("2017", periods=4, tz="US/Central"))

~/sandbox/pandas-ip/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    239                                        raise_cast_failure=True)
    240
--> 241                 data = SingleBlockManager(data, index, fastpath=True)
    242
    243         generic.NDFrame.__init__(self, data, fastpath=True)

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, block, axis, do_integrity_check, fastpath)
   4467
   4468         if not isinstance(block, Block):
-> 4469             block = make_block(block, placement=slice(0, len(axis)), ndim=1)
   4470
   4471         self.blocks = [block]

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in make_block(values, placement, klass, ndim, dtype, fastpath)
   3006                      placement=placement, dtype=dtype)
   3007
-> 3008     return klass(values, ndim=ndim, placement=placement)
   3009
   3010 # TODO: flexible with index=None and/or items=None

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim, dtype)
   2662
   2663         super(DatetimeTZBlock, self).__init__(values, placement=placement,
-> 2664                                               ndim=ndim)
   2665
   2666     def copy(self, deep=True, mgr=None):

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim)
   1700                 ndim = 2
   1701         super(NonConsolidatableMixIn, self).__init__(values, placement,
-> 1702                                                      ndim=ndim)
   1703
   1704     @property

~/sandbox/pandas-ip/pandas/pandas/core/internals.py in __init__(self, values, placement, ndim)
   2514         if values.dtype != _NS_DTYPE:
   2515             # not datetime64 or datetime64tz
-> 2516             values = conversion.ensure_datetime64ns(values)
   2517
   2518         super(DatetimeBlock, self).__init__(values,

TypeError: Argument 'arr' has incorrect type (expected numpy.ndarray, got DatetimeIndex)
```

jreback replied:
something is wrong if you have to change this.

jorisvandenbossche replied:
As Tom said before, this is because DatetimeTZBlock also calls this init method, and they have a `DatetimeIndex` as holder. So this check needs to let a DatetimeIndex through (and not only a datetime64 dtyped array).  
Tom, is that a correct interpretation? In any case a long comment to clarify could be useful.

TomAugspurger replied:
How about a method on `DatetimeBlock` and `DatetimeTZBlock` that checks this?

For `DatetimeBlock`:

```
def maybe_ensure_datetime64ns(self, values):
    if values.dtype != _NS_DTYPE:
```

And in `DatetimeTZBlock`:

```
def maybe_ensure_datetime64ns(self, values):
    if values.dtype.base != _NS_DTYPE:
```

That's more indirection, but more correct?

TomAugspurger replied:
@jorisvandenbossche the `__init__` change is the root issue, yes.

I think defining a method on each block, like `_maybe_coerce_values` or something, may be a useful refactoring anyway. Lot's of subclasses are essentially doing this in the `__init__` anyway, so we could remove all those and just call `_maybe_coerce_values` in the parent.

Will do for Datetime now, and we can expand if we like it.

jorisvandenbossche replied:
In any case, it would be more correct. As with the change above you allow datetime tz values also for a DatetimeBlock, which should not be accepted. So I think that would be a good idea to have the check in a method that can be overridden by DatetimeTZBlock

TomAugspurger replied:
Sorry, confused myself again, it *is* doing coercion. We have I was looking at `datetimes._ensure_datetime64`, which does validation, not `conversion.ensure_datetime64ns`, which does coercion.

TomAugspurger replied:
The `dtype` argument to `DatetimeTZBlock`  makes this difficult, since no other block accepts it. If we can remove that and do the conversion at the caller, it'd be a nice cleanup. Going to try to work around it for now though and do in a followup.

jreback reviewed (changes requested):
I would rather start with a minimal implementation of ExtensionBlock. Then fix what needs to be fixed, rather than simply move it to EB. 

We need quite a bit of cleanup in internals in order to do this cleanly. Certainly can move it in this direction. But would like to see incremental PR's to clean up first.

jorisvandenbossche commented:
> I would rather start with a minimal implementation of ExtensionBlock. Then fix what needs to be fixed, rather than simply move it to EB.

@TomAugspurger will be able to give more details, but I think that this is actually what he has been doing. Adding things to ExtensionBlock to make things working.

shoyer commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+            * A scalar integer position
+            * A slice object, where 'start', 'stop', and 'step' are
+              integers or None
+            * A 1-d boolean NumPy ndarray the same length as 'self'
```
We should be strict about how we use this interface from pandas. I see little harm in saying this is always a NumPy array since we are the only ones calling it and some clear advantages. Otherwise, how is a user supposed to detect this case?

TomAugspurger committed:
Handle get_values, to_dense, is_view

TomAugspurger commented:
> @TomAugspurger will be able to give more details, but I think that this is actually what he has been doing. Adding things to ExtensionBlock to make things working.

Correct.

TomAugspurger committed:
Docs

TomAugspurger committed:
Remove is_extension, is_bool

Remove inherited convert

TomAugspurger committed:
Sparse formatter

TomAugspurger commented on pandas/io/formats/format.py:
```diff
@@ -2115,6 +2118,15 @@ def _format_strings(self):
 
+class SparseArrayFormatter(GenericArrayFormatter):
+
+    def _format_strings(self):
```
There was an issue with NumPy 1.10 and masking for sparse arrays. Seeing if this passes our CI quick, but I should be able to remove it.

edit: yep, everything passes. Let me see if I can simplify.

TomAugspurger committed:
Revert "Sparse formatter"

This reverts commit ab2f0457839fece3b3ef067f29994b42908bd037.

TomAugspurger committed:
Unbox SparseSeries

TomAugspurger commented:
For reference, the changes in https://github.com/pandas-dev/pandas/pull/19268#pullrequestreview-92638262 turned up an issue in master with `SparseArray` blocks being consolidated,

On master:

```python
In [1]: import pandas as pd; import numpy as np

In [2]: sdf = pd.SparseDataFrame([[np.nan, 1], [2, np.nan]])

In [3]: sdf._data.blocks
Out[3]:
(SparseBlock: slice(0, 1, 1), 1 x 2, dtype: float64,
 SparseBlock: slice(1, 2, 1), 1 x 2, dtype: float64)

In [4]: sdf[0][1] = 0
/Users/taugspurger/Envs/pandas-dev/bin/ipython:1: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3.6

In [5]: sdf._data.blocks
Out[5]: (SparseBlock: slice(1, 2, 1), 1 x 2, dtype: float64,)
```

On this branch the blocks are (correctly) not consolidated. Adding a test to confirm.

TomAugspurger committed:
Added test for sparse consolidation

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,191 @@
+    Notes
+    -----
+    pandas will recognize instances of this class as proper arrays
+    with a custom type and will not attempt to coerce them to objects.
```
Added a bit about 1-D and some high-level examples.

TomAugspurger committed:
Docs

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into pandas-array-interface-3

TomAugspurger commented:
Looking into performance things now. `isinstance` checks are a decent amount slower

```python
s = pd.Series()
%timeit isinstance(s, pd.api.types.CategoricalDtype)
```

Master: `281 ns ± 7.39 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)`
Branch: `25.5 µs ± 4.23 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)`

This add up since we do a lot of `isinstance` calls via the `is_*` methods.

Looking into overriding `__instancecheck__` without breaking things.

jbrockmendel commented:
w/r/t isinstance checks there's an abandoned #17599 PR that moved the ABCFoos to cython, got a boost.  IIRC the existing `isinstance(item, ABCFoo)` check is about 2x slower than `isinstance(item, Foo)`.

TomAugspurger commented:
This is for python's `abc.ABCMeta`, which seems to be 100x slower than that even.

TomAugspurger committed:
Moved to errors

TomAugspurger committed:
Handle classmethods, properties

TomAugspurger committed:
Use our AbstractMethodError

TomAugspurger commented:
- Moved AbstractMethodError to `pandas.errors` in https://github.com/pandas-dev/pandas/pull/19268/commits/fc688a56e20e62e6b7b806aedc4a4a760af079b1 to avoid a circular depedency
- Modified AbstractMethodError to have proper error messages for classmethods and porperties in https://github.com/pandas-dev/pandas/pull/19268/commits/fbc846644be3e97314b203dd3eab9a36ef50a274
- Using AbstrctMethodError instead of `abc.ABCMeta` in https://github.com/pandas-dev/pandas/pull/19268/commits/030bb194e523b24442db77c8cbabe16315880c73

Rerunning the ASV now, but I think this initial PR ready.

shoyer commented:
You can always use "type(...) is" if necessary. That avoids looking through
the class hierarchy so it's generally must faster.

On Wed, Jan 31, 2018 at 8:52 AM Tom Augspurger <notifications@github.com>
wrote:

> Looking into performance things now. isinstance checks are a decent
> amount slower
>
> s = pd.Series()%timeit isinstance(s, pd.api.types.CategoricalDtype)
>
> Master: 281 ns ± 7.39 ns per loop (mean ± std. dev. of 7 runs, 1000000
> loops each)
> Branch: 25.5 µs ± 4.23 µs per loop (mean ± std. dev. of 7 runs, 10000
> loops each)
>
> This add up since we do a lot of isinstance calls via the is_* methods.
>
> Looking into overriding __instancecheck__ without breaking things.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19268#issuecomment-361995800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1nTZA6skY30TN5O3KhKdIl2UMG9Oks5tQJo4gaJpZM4Rf8hq>
> .
>

TomAugspurger committed:
Lint

TomAugspurger commented:
ASV is good now that we aren't using `ABCMeta`. Personally, I'm OK with that. We can provide base classes to test against like https://github.com/TomAugspurger/pandas/blob/pandas-array-upstream%2Binterval/pandas/tests/extension_arrays/base.py if library authors are interested.

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,238 @@
+    # ------------------------------------------------------------------------
+    def isna(self):
+        # type: () -> np.ndarray
+        """Boolean NumPy array indicating if each value is missing."""
```
same length as self

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,238 @@
+    # Block-related methods
+    # ------------------------------------------------------------------------
+    @property
+    def _fill_value(self):
```
need to list this in the very top doc-string

jreback commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,238 @@
+        """
+        raise AbstractMethodError(cls)
+
+    def _can_hold_na(self):
```
need to list this in the very top doc-string

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -2130,6 +2132,20 @@ def repeat(self, repeats, *args, **kwargs):
         return self._constructor(values=codes, categories=self.categories,
                                  ordered=self.ordered, fastpath=True)
 
+    # ExtensionArray Interface things
```
maybe compat for ExtensionArray

jreback commented on pandas/core/dtypes/base.py:
```diff
@@ -0,0 +1,125 @@
+
+    @property
+    def names(self):
+        # type: () -> Optional[List[str]]
```
add an example here (or maybe more expl)

jreback commented on pandas/core/internals.py:
```diff
@@ -1674,24 +1680,19 @@ class NonConsolidatableMixIn(object):
     _holder = None
 
     def __init__(self, values, placement, ndim=None):
+        # Placement must be converted to BlockPlacement so that we can check
```
can you make a doc-string here

jreback commented on pandas/core/internals.py:
```diff
@@ -1800,6 +1801,91 @@ def _unstack(self, unstacker_func, new_columns):
+
+    Notes
+    -----
+    This is the holds all 3rd-party extension types. It's also the immediate
```
typo

jreback commented on pandas/core/internals.py:
```diff
@@ -1800,6 +1801,91 @@ def _unstack(self, unstacker_func, new_columns):
+        return self.make_block_same_class(new_values, new_mgr_locs)
+
+    def _can_hold_element(self, element):
+        # XXX: We may need to think about pushing this onto the array.
```
can you add some doc-strings

jreback commented on pandas/core/internals.py:
```diff
@@ -2734,6 +2797,13 @@ class SparseBlock(NonConsolidatableMixIn, Block):
     _concatenator = staticmethod(_concat._concat_sparse)
 
+    def __init__(self, values, placement, ndim=None):
+        # Ensure that we have the underlying SparseArray here...
```
so this should be on ExtensionArray. I know that Sparse is not there yet, but the idea is that you should call a method to do this coercion, IOW so each sub-class doesn't have to repeat it.

TomAugspurger replied:
Agreed. We should be able to remove the `__init__` for `DatetimeBlock`, `CategoricalBlock`, and `SparseBlock`.

jreback commented on pandas/tests/dtypes/test_dtypes.py:
```diff
@@ -742,3 +744,31 @@ def test_categorical_categories(self):
+        pd.Categorical([]).dtype,
+        pd.Series(pd.Categorical([])),
+        DummyDtype(),
+        DummyArray(),
```
can you test all of the exposed methods / properties

TomAugspurger replied:
The only thing I could test on the dummy ones is `kind`, which returns just 'O' and `names`, which I'm hopeful we can remove.

All the other methods are abstract or rely on an abstract method.

I do have real classes that implement the interface, but waiting for the interface to settle a bit before rewriting those classes and the tests.

jreback commented on pandas/tests/internals/test_external_block.py:
```diff
@@ -5,12 +5,12 @@
 
 import pandas as pd
 from pandas.core.internals import (
```
ideally some more tests

TomAugspurger replied:
Many more tests in the followup PR, coming as soon as this is merged: https://github.com/TomAugspurger/pandas/compare/pandas-array-interface-3...TomAugspurger:pandas-array-upstream+interval#diff-1837f9e869f48877d9a16e2901942b60

Anything in particular you want tested at the block level? Most of those are at the NDFrame level.

jreback reviewed (changes requested):
looks pretty good. some comments and questions.

TomAugspurger committed:
Cleanup

TomAugspurger committed:
Move ndim validation to a method.

jorisvandenbossche commented:
The practical consequence of no longer using abc.ABCMeta, is it only that you now cannot register a class but need to actually subclass? (which doesn't seem a big problem?)

TomAugspurger commented:
That, and library author-wise, it's now a bit harder to validate that your extension array implements everything. Now it's possible to instantiate an `ExtensionArray` that doesn't implement the interface. With an ABC that would raise an exception.

jreback commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -5,15 +5,15 @@
     THIS IS NOT A REAL NUMPY DTYPE
     """
-    name = None
-    names = None
```
can you add some docs about these attributes

TomAugspurger replied:
They're documented in `ExtensionDtype`.

jreback commented on pandas/core/internals.py:
```diff
@@ -109,18 +110,31 @@ class Block(PandasObject):
                 'Wrong number of items passed {val}, placement implies '
                 '{mgr}'.format(val=len(self.values), mgr=len(self.mgr_locs)))
 
+    def _maybe_validate_ndim(self, values, ndim):
```
thanks. much nicer.

jreback commented on pandas/core/internals.py:
```diff
@@ -1800,6 +1820,91 @@ def _unstack(self, unstacker_func, new_columns):
+    ExtensionArrays are limited to 1-D.
+    """
+    def __init__(self, values, placement, ndim=None):
+        self._holder = type(values)
```
I think _holder can/should be a property of the Block itself (it can be a cached property)

TomAugspurger replied:
`Block` defines `_holder` as a class attribute. Why would caching it be helpful? It's not reused among multiple `ExtensionBlocks`.

jreback replied:
its just 1 more thing to do in the __init__ which is not needed

TomAugspurger committed:
Try this

TomAugspurger committed:
Make ExtensionBlock._holder a property

Removed ExtensionBlock.__init__

TomAugspurger committed:
Make _holder a property for all

TomAugspurger commented:
Done in
https://github.com/pandas-dev/pandas/pull/19268/commits/cd0997e354121bc0414ef1675cb5b8241944a9b0
if you could take a look.

I did it for every block, which seemed better than a mix of attributes and
properties.



On Thu, Feb 1, 2018 at 1:07 PM, Jeff Reback <notifications@github.com>
wrote:

> *@jreback* commented on this pull request.
> ------------------------------
>
> In pandas/core/internals.py
> <https://github.com/pandas-dev/pandas/pull/19268#discussion_r165349895>:
>
> > @@ -1800,6 +1820,91 @@ def _unstack(self, unstacker_func, new_columns):
>          return blocks, mask
>
>
> +class ExtensionBlock(NonConsolidatableMixIn, Block):
> +    """Block for holding extension types.
> +
> +    Notes
> +    -----
> +    This holds all 3rd-party extension array types. It's also the immediate
> +    parent class for our internal extension types' blocks, CategoricalBlock.
> +
> +    ExtensionArrays are limited to 1-D.
> +    """
> +    def __init__(self, values, placement, ndim=None):
> +        self._holder = type(values)
>
> its just 1 more thing to do in the *init* which is not needed
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19268#discussion_r165349895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIoHyKRRVAeZqotWc_9Qq56iVKPs_ks5tQbctgaJpZM4Rf8hq>
> .
>

TomAugspurger committed:
Refactored validate_ndim

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -108,12 +108,7 @@ class Block(PandasObject):
     _concatenator = staticmethod(np.concatenate)
 
     def __init__(self, values, placement, ndim=None):
-        if ndim is None:
```
cc @jbrockmendel. Pretty sure this is what you had in mind. Agreed it's cleaner.

TomAugspurger replied:
(needed a followup commit to use `self.ndim` a few lines down.

TomAugspurger committed:
fixup! Refactored validate_ndim

TomAugspurger committed:
lint

TomAugspurger commented:
CI is all green.

TomAugspurger commented:
Just to make sure, @jorisvandenbossche, @shoyer are you +1 on the changes here?

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,247 @@
+    * take
+    * copy
+    * _formatting_values
+    * _concat_same_type
```
The two lines above can be removed here and mentioned in the list below (actually only formatting_values, as concat_same_type is already there)

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,247 @@
+    an instance, not error.
+
+    Additionally, certain methods and interfaces are required for proper
+    this array to be properly stored inside a ``DataFrame`` or ``Series``.
```
this is repetitive with above (the list of methods that are required), and there is also a typo in "for ~~proper~~ this array to be properly"

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,247 @@
+
+        Examples
+        --------
+        Suppose the extension array somehow backed by a NumPy structured array
```
let's make this just "NumPy array", as this is not specific to structured arrays

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,247 @@
+           def take(self, indexer, allow_fill=True, fill_value=None):
+               mask = indexer == -1
+               result = self.data.take(indexer)
+               result[mask] = self._fill_value
```
One question here is: should the keyword argument `fill_value` actually be honored? (like `if fill_value is None: fill_value = self._fill_value`) 
Are there case where pandas will actually pass a certain value? 

In any case some clarification would be helpful, also if it is just in the signature for compatibility but may be ignored (maybe in a follow-up).

TomAugspurger replied:
re `fill_value`, I based this off `Categorical.take`. It does `assert isna(fill_value)`, but otherwise ignores it.

I think that since ExtensionArray.take returns an ExtensionArray, most implementations will just ignore fill_value. I'll clarify it in the docs.

jorisvandenbossche commented on pandas/core/arrays/base.py:
```diff
@@ -0,0 +1,247 @@
+        # type: () -> np.ndarray
+        # At the moment, this has to be an array since we use result.dtype
+        """An array of values to be printed in, e.g. the Series repr"""
+        raise AbstractMethodError(self)
```
Maybe we can provide a default implementation of `return np.asarray(self)` ? (so a densified object array). That is what I do in geopandas, and I suppose would also work for the IPadresses ?

TomAugspurger replied:
Yes, I suppose that'll be OK for many implementations.

jorisvandenbossche reviewed (approved):
Yes, +1 to merge this now, to be able to proceed with the other parts. We will probably need to make minor tweaks to the interface, but that will only become clear by doing next PRs to actually use this.

Added some minor doc comments. And two comments on the interface (take fill_value kwarg and formatting_values), but that can also go in a follow-up.

TomAugspurger commented:
Thanks @jorisvandenbossche, I've fixed up the docs in my followup branch.

shoyer commented:
Yes, I'm pretty sure I'm +1 on the current state, though right now GitHub
is crashing when I try to load the page with this PR :).

On Fri, Feb 2, 2018 at 11:13 AM Tom Augspurger <notifications@github.com>
wrote:

> Just to make sure, @jorisvandenbossche
> <https://github.com/jorisvandenbossche>, @shoyer
> <https://github.com/shoyer> are you +1 on the changes here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19268#issuecomment-362678155>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1pAcUdeZ9GkAMr2bUAfuH7Fs0hZeks5tQ15DgaJpZM4Rf8hq>
> .
>

TomAugspurger commented:
> Yes, I'm pretty sure I'm +1 on the current state, though right now GitHub
is crashing when I try to load the page with this PR :).

This is partly why I'm hoping to push further changes into a followup :)

jorisvandenbossche commented:
Yes, for me the same. I think the page is only loading in 1 out of 5 tries
or so .. Another reason to merge :-)

2018-02-02 20:57 GMT+01:00 Stephan Hoyer <notifications@github.com>:

> Yes, I'm pretty sure I'm +1 on the current state, though right now GitHub
> is crashing when I try to load the page with this PR :).
>
> On Fri, Feb 2, 2018 at 11:13 AM Tom Augspurger <notifications@github.com>
> wrote:
>
> > Just to make sure, @jorisvandenbossche
> > <https://github.com/jorisvandenbossche>, @shoyer
> > <https://github.com/shoyer> are you +1 on the changes here?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/pandas-dev/pandas/pull/19268#issuecomment-362678155
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ABKS1pAcUdeZ9GkAMr2bUAfuH7Fs0hZeks5tQ15DgaJpZM4Rf8hq>
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19268#issuecomment-362690105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-SUJgmb4nGDrSBl0URDRGzsEeR9p_8ks5tQ2iPgaJpZM4Rf8hq>
> .
>

jreback commented:
i haven’t had a look
but i guess it’s ok

TomAugspurger commented:
Alrighty, thanks! Followup PRs inbound :)

jorisvandenbossche commented:
> i haven’t had a look
but i guess it’s ok

don't hesitate to still add comments here, they can then always be dealt
with in follow-up prs

2018-02-02 22:34 GMT+01:00 Tom Augspurger <notifications@github.com>:

> Alrighty, thanks! Followup PRs inbound :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19268#issuecomment-362714327>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-SUMR-MY81ioq-2REvrqJaRP6eN4BAks5tQ39zgaJpZM4Rf8hq>
> .
>
